units/provisioning_setup_host.go

```diff
@@ -898,3 +916,98 @@ func (j *setupHostJob) tryRequeue(ctx context.Context) {
 func shouldRetryProvisioning(h *host.Host) bool {
 	return h.ProvisionAttempts <= provisionRetryLimit && h.Status == evergreen.HostProvisioning && !h.Provisioned
 }
+
+func attachVolume(ctx context.Context, env evergreen.Environment, h *host.Host) error {
+	if !(h.Distro.JasperCommunication() && h.Distro.IsLinux()) {
+		return errors.Errorf("host '%s' of distro '%s' doesn't support mounting a volume", h.Id, h.Distro.Id)
+	}
+
+	if h.HomeVolumeID == "" {
+		mgrOpts, err := cloud.GetManagerOptions(h.Distro)
+		if err != nil {
+			return errors.Wrapf(err, "can't get ManagerOpts for '%s'", h.Id)
+		}
+		cloudMgr, err := cloud.GetManager(ctx, env, mgrOpts)
+		if err != nil {
+			return errors.Wrapf(err,
+				"failed to get cloud manager for host %s with provider %s",
+				h.Id, h.Provider)
+		}
+
+		// create the volume
+		volume, err := cloudMgr.CreateVolume(ctx, &host.Volume{
+			Size:             h.HomeVolumeGB,
+			AvailabilityZone: h.Zone,
+			CreatedBy:        h.StartedBy,
+			Type:             evergreen.DefaultEBSType,
+		})
+		if err != nil {
+			return errors.Wrapf(err, "can't create a new volume for host '%s'", h.Id)
+		}
+
+		// attach to the host
+		attachment := host.VolumeAttachment{VolumeID: volume.ID}
+		if err = cloudMgr.AttachVolume(ctx, h, &attachment); err != nil {
+			return errors.Wrapf(err, "can't attach volume '%s' to host '%s'", volume.ID, h.Id)
+		}
+
+		if err = h.SetHomeVolume(volume.ID, attachment.DeviceName); err != nil {
+			return errors.Wrapf(err, "can't set home volume for host '%s'", h.Id)
+		}
+	}
+
+	// run the distro's mount script
+	var err error
+	for i := 0; i < mountRetryLimit; i++ {
+		if err = errors.Wrap(mountLinuxVolume(ctx, env, h), "can't mount volume"); err == nil {
+			return nil
+		}
+		time.Sleep(mountSleepDuration)
+	}
+	return err
+}
+
+func mountLinuxVolume(ctx context.Context, env evergreen.Environment, h *host.Host) error {
+	client, err := h.JasperClient(ctx, env)
+	if err != nil {
+		return errors.Wrap(err, "can't get Jasper client")
+	}
+	defer func() {
+		grip.Warning(message.WrapError(client.CloseConnection(), message.Fields{
+			"message": "could not close connection to Jasper",
+			"host":    h.Id,
+			"distro":  h.Distro.Id,
+		}))
+	}()
+
+	deviceName := h.HomeVolumeDeviceName
+	if h.Distro.HomeVolumeDeviceName != "" {
+		deviceName = h.Distro.HomeVolumeDeviceName
+	}
+
+	// continue on umount mount error
+	exitCode, err := host.RunJasperCommandSynchronously(ctx, client, &options.Create{Args: []string{"sudo", "umount", "-f", fmt.Sprintf("/dev/%s", deviceName)}})
+	if err != nil && exitCode != umountMountErrorCode {
+		return errors.Wrap(err, "problem running umount command")
+	}
+
+	commands := []options.Create{
+		{Args: []string{"sudo", "/sbin/mkfs.xfs", "-f", fmt.Sprintf("/dev/%s", deviceName)}},
```

Comment:
I worry about hardcoding the filesystem type here? this seems like it should be distro settable?

Reply:
done.
I made the entire command configurable (e.g. `/sbin/mkfs.xfs -f`) because the various formatting utilities take different arguments.